File: app/widgets/library/citations/authors_catalog.js
//= require <templates/catalog>
//= require <templates/null>
//= require <library/citations/authors_catalog_action_panel>
//= require <simple_selected>

JooseModule('Library.Citations', function () {
  JooseClass('AuthorsCatalog', {
    isa: Templates.Catalog,
    has: {
      authors: { is: 'ro', required: true, nullable: false },
      collection: { is: 'ro', init: function () { return this.authors() } },
      columns: { is: 'ro', init: function () { return [
        { attribute: 'name', label: 'Author', width: 300 },
        { moveControls: true, label: '', width: 100 }
      ]
      } },
      limit: { is: 'ro', init: null },
      hasContentsForm: { is: 'ro', init: false },
      hasFilterSet: { is: 'ro', init: false },
      selected: { is: 'ro',
        lazy: true, // lazy required, this.collection() does not exist yet
        init: function () { return new SimpleSelected({ catalog: this }) }
      },
      widgets: { is: 'ro', init: function () { return $Reg({
        actionPanel: new Library.Citations.AuthorsCatalogActionPanel({ parent: this }),
        filterSet:   new Templates.Null({ parent: this })
      }, this) } }
    }
  });
});

File: app/widgets/library/citations/authors_catalog.rb
class Library::Citations::AuthorsCatalog < Templates::Catalog
  def initialize options = {}
    @authors ||= options[:authors] || fail('authors required')
    @collection ||= @authors
    @parent ||= options[:parent]
    @context ||= options[:context] || parent.context
    mode = interact_mode == 'browse' ?  false : true
    @columns ||= [
      { attribute: :name, label: 'Author', width: 300, order_on: false },
      { move_controls: mode, label: '', width: 100 , order_on: false  }
    ]
    @limit ||= nil
    @has_contents_form ||= false
    @has_filter_set ||= false
    widgets({
      action_panel: { init: ->{ Library::Citations::AuthorsCatalogActionPanel.new({ parent: self }) }},
      filter_set:   { init: ->{ Templates::Null.new({parent: self }) }}
    })
    super
  end

  alias_method :to_s, :render_to_string
end
